scalameta/common/shared/src/main/scala/scala/meta/internal/trees/ast.scala

```diff
@@ -336,7 +350,39 @@ class AstNamerMacros(val c: Context) extends AstReflection with CommonNamerMacro
           }
           internal(...$internalArgss)
         }
-      """
+        """
+
+        // step 13a: generate additional binary compat Companion.apply
+        // generate all possible applies so with field A, B and additional binary compat ones C and D, we generate:
+        // apply(C, D, A), apply(C, D, B), apply(C, D, A, B)
+        val sortedBinaryCompatName = binaryCompatVars
+          .collect { case vr: ValDef => vr }
+          .sortBy(_.name.toString())
+        1.until(binaryCompatNum + 1).foreach { size =>
+          sortedBinaryCompatName.combinations(size).foreach { fields =>
```

Comment:
Will this compile successfully if you have two or more binary compact fields with the same type?

```scala
@ast class Template(a: A, c: C) {
  @bincompat var b1: B
  @bincompat var b2: B
  @bincompat var b3: B
}
// generates
def apply(a: A, c: C, b1: B)
def apply(a: A, c: C, b2: B) // Error: same erasure as previous method
def apply(a: A, c: C, b3: B) // Error: same erasure as previous method
def apply(a: A, c: C, b1: B, b2: B)
...
```

To support adding new bincompat fields without breaking changes, I think we need to generate a new `apply` method for every bincompat field.

```
def apply(a: A, c: C, b1: B)
def apply(a: A, c: C, b1: B, b2: B)
def apply(a: A, c: C, b1: B, b2: B, b3: B)
```

This requires the ordering of the `apply` methods to match the declaration order. Otherwise, we would have to make sure to name new `@bincompat` fields to appear last in lexicographical order.

Reply:
Every new bincompat field should have the equivalent of a `@since` annotation from https://github.com/alexarchambault/data-class#adding-fields

Reply:
Ach, you're right. I will need to just generate a single new apply for each new added field. Should be a small change. I went overboard with the combinations.

Not sure if we need `since` here, it's needed when we add multiple fields at the same time and we want a single new apply. I think we should be fine with adding one new apply for each new compat field.  What do you think?

Reply:
Ach no, you are right. We need something like since.

Reply:
@olafurpg I added `since` field to `@binaryCompatField` and order the fields using it. I also switched to just generating one new apply per field.

Reply:
If you implemented `@since` then that's great! I was thinking that we would infer `@since` for every bincompat field. I suspect it will be rare that we add two bincompat fields and don't want to add two new apply methods.

Reply:
It's not really `@since`. It will still add two new applies. The only problem I see is if we add two new fields at the same version in different PRs and add the new one before the previous one.  Although I expect `@since` doesn't work in that case either :sweat_smile: